### ISSUE TITLE
Orientation issues

### ISSUE THREAD

**Issue by Cody-G** [JuliaIO/NRRD.jl#17]
I have an image for which I write an `AxisArray` as an NRRD and I get a header like this:

```
NRRD0004
type: float
dimension: 3
space dimension: 3
sizes: 390 256 61
spacings: 1.3 1.3 3.0
units: "μm" "μm" "μm"
endian: little
encoding: raw
space origin: (0.32499999999998863,0.32500000000001705,0.0)
```

When I try loading this image with [ITK](https://itk.org/) it fails with this message:
```
itk::ExceptionObject (0x56a9460)
Location: "unknown" 
File: /home/cody/src/antsbin/ITKv4/Modules/IO/NRRD/src/itkNrrdImageIO.cxx
Line: 474
Description: itk::ERROR: NrrdImageIO(0x56a8ae0): ReadImageInformation: Error interpreting nrrd spacing (nrrdSpacingStatusScalarWithSpace)
```

If I modify the header in this way it loads fine (new fields are starred):
```
NRRD0004
type: float
dimension: 3
space dimension: 3
sizes: 390 256 61
**space units: "μm" "μm" "μm"**
endian: little
encoding: raw
space origin: (0.32499999999998863,0.32500000000001705,0.0)
**space directions: (1.3,0,0) (0,1.3,0) (0,0,3.0)**
```

I'm a little confused about why this change was necessary.  When reading the NRRD format spec it seems that the "spacings" field should be equivalent to the "space directions" field when the world axes align with the array axes, as in this case.  Likewise I don't understand why we need to use the "space units" field instead of the "units" field, but both changes are necessary.

I can submit a PR changing how we write the header.  But first I'd like to make sure that I'm handling this in the best way.  Any opinions from anyone more familiar with the NRRD format?

**Comment by timholy** [comment]
I'm pretty unclear on whether this is necessary, too. Bringing in the heavy artillery: @kindlmann?

**Comment by kindlmann** [comment]
Sorry for the slow reply.  Thank you for pinging me; I'm happy to explain this stuff.

There are basically two distinct regimes for representing array orientation in NRRD.  (1) Using the "space" fields, or (2) a confusing mix of other fields that can only represent axis-aligned grids.  They should not be mixed.  In my own work, I've been trying to push things more towards using the "space" fields, because it offers richer semantics for identifying what the world space actually is.

In your first header, you're trying to both assert something about the location of the first sample in world-space (with "space origin" and "space dimension"), but also some axis-aligned spacing information (with "spacings").  I *think* "unu" would be ok with this header, but ITK's handling of array orientation is quite simple compared to NRRD's, so certain restrictions are enforced before things get transferred from NRRD to ITK.  As far as I recall (when I was helping write the ITK<-->NRRD code about a decade ago) we were very interested in minimizing changes to the orientation semantics in ITK.

In your second header, now you're being explicit about you move in space (the same 3D space in which the origin lives) as you move to subsequent samples along each of the three axes.  Notice that while

`space directions: (1.3,0,0) (0,1.3,0) (0,0,3.0)
`

works, so do 

```
space directions: (0,1.3,0) (0,0,1.3) (3.0,0,0)
space directions: (0,0,1.3) (1.3,0,0) (0,3.0,0)
space directions: (-1.3,0,0) (0,1.3,0) (0,0,3.0)
space directions: (0,-1.3,0) (0,0,1.3) (3.0,0,0)
space directions: (0,0,-1.3) (1.3,0,0) (0,3.0,0)
space directions: (1.3,0,0) (0,-1.3,0) (0,0,3.0)
space directions: (0,1.3,0) (0,0,-1.3) (3.0,0,0)
space directions: (0,0,1.3) (-1.3,0,0) (0,3.0,0)
```

and so on: there are many different array orientations that can be described with "spacings between samples are 1.3, 1.3, 3.0", half of which will have implications for the handed-ness of any structures in your data.  The "space directions" fields invite you to figure out exactly what you intend, as well as allowing representations of more general orientations, with rotations, and shears (as from gantry tilt on CT scans, or accounting for per-slice time-of-acquisition 3D images over time)

Some of these semantics are moot if the world space itself is identified only by "space dimension: 3".  Is this a right-handed or a left-handed world-space coordinate system?  The fields like "space: LPS" or "space: 3D-right-handed" provide a way of documenting this.

If the idea of world-space and all the semantics subtleties seem tiresome or irrelevant, then you can also have a header like:

```
NRRD0004
type: float
dimension: 3
sizes: 390 256 61
spacings: 1.3 1.3 3.0
axis mins: 0.32499999999998863 0.32500000000001705 0.0
units: "μm" "μm" "μm"
endian: little
encoding: raw

```

This converts your "origin" to a set of per-axis minimums; that's also a valid NRRD header.  However, without specifying if this is a cell-centered vs a node-centered sampling, there could be some ambiguity about where the first sample is relative to those per-axis minimums.

That was probably more than you wanted to know!  Let me know if I can help clarify or muddy anything else.

**Comment by timholy** [comment]
Thanks @kindlmann, that was really helpful. I appreciate this level of detail.

While I have your attention, may I ask you about something we noticed that led me to design an "extension" to NRRD. The issue that triggered this was owning a scientific camera that produced 14-bit values, stored in a `uint16` with the top two bits 0. Since being able to detect saturation is pretty important (and the naive way, based on the `uint16` data format, would be to compare to 65535 which would be wrong), it pointed out the need to encode the dynamic range of the sensor somehow.

I came up with [this strategy](https://github.com/JuliaIO/NRRD.jl/blob/7a052b65c832728a762ae0b178739dd9fcfdf7ce/src/NRRD.jl#L478-L521). I'd be interested in feedback. You can ignore the part about `FixedPoint` numbers (or refer to http://juliaimages.github.io/latest/quickstart.html#Colors,-the-0-to-1-intensity-scale,-and-views-1).

**Comment by kindlmann** [comment]
Interesting.  To make sure I understand: is this addressing the problem of communicating something about the imaging modality (and whatever pre-processing) produced the array, namely how the max value (or white) is represented, which is related to but not entirely determined by how (with which type) the values themselves are represented?

That is not something I'd previously thought about representing. I wonder if we should add a new field "sample max" to store this (I'm happy to do that; there are some new spaces and encodings that have already necessitated creating a new NRRD0006 magic).  In the mean time, I think hijacking "sample units" is probably appropriate. "Sample units" is, however, supposed to be things like "meters" or "feet" (for elevation data) or "watts/m^2" (for irradiance).  (when designing the format I spent a few days learning about different standards for representing and encoding scientific units, and eventually punted and said: "screw it: strings").

What I don't understand is the interaction between this two-part string and array shape (dimension, and #samples on each axis), the "kinds" label of some axis, and the sample type.  Does "space units: rgb (1.0,1.0,1.0)" mean that there is some length-three axis axis, and that is the one that should be be interpreted as the rgb axis?  What if there are two length-three axes?  Does one of them have to have a matching kind ("rgb")? What happens to this field if we slice the array along the color axis to get out the green channel (should sample units be updated to "gray 1.0"?).  Is the "space units: rgb (1.0,1.0,1.0)" valid regardless of sample type (uchar or float)?  Does it need updating if floating point values are quantized to 16 bits (in which case does it interact with the oddly-named "old min" and "old max" fields?)?

The number of questions indicate my level of interest, and not any judgement.

**Comment by timholy** [comment]
> I wonder if we should add a new field "sample max"

I'd be very happy if this were a new field; I chose "sample units" simply because it was the closest analog I could find among the existing fields. While a single number would surely suffice for most uses, there are cases where it's insufficient; see the two HSV examples in that documentation string. Currently there doesn't appear to be any way to know which of the two HSV conventions is being employed, and I could cook up an artificial file in which all of the values would be valid with either convention but have vastly different meanings.

>  (when designing the format I spent a few days learning about different standards for representing and encoding scientific units, and eventually punted and said: "screw it: strings").

Yes, it's not a simple subject. Here we use Julia's amazing Unitful.jl package to [convert some of these into things you can do math with](https://github.com/JuliaIO/NRRD.jl/blob/7a052b65c832728a762ae0b178739dd9fcfdf7ce/src/NRRD.jl#L108-L110), but obviously we recognize only a subset of the strings that users might supply.

> Does "space units: rgb (1.0,1.0,1.0)" mean that there is some length-three axis axis, and that is the one that should be be interpreted as the rgb axis? What if there are two length-three axes?

Unsurprisingly, you ask all the right questions. Here's a demo:
```julia
julia> using FileIO

julia> open("/tmp/test.nrrd", "w") do io
           print(io, """
       NRRD0004
       type: uint8
       dimension: 4
       sizes: 3 3 3 3
       kinds: RGB-color space space time
       endian: little
       encoding: raw
       sample units: rgb (255,255,255)

       """)
           write(io, rand(UInt8, 3, 3, 3, 3))
       end
81

julia> img = load("/tmp/test.nrrd")
3-dimensional AxisArray{RGB{N0f8},3,...} with axes:
    :space_1, 0:2
    :space_2, 0:2
    :time, 0:2
And data, a 3×3×3 Array{RGB{N0f8},3}:
[:, :, 1] =
 RGB{N0f8}(0.612,0.075,0.302)  RGB{N0f8}(0.114,0.173,0.533)  RGB{N0f8}(0.263,0.769,0.396)
 RGB{N0f8}(0.855,0.408,0.722)  RGB{N0f8}(0.337,0.122,0.118)  RGB{N0f8}(0.863,0.886,0.459)
 RGB{N0f8}(0.369,0.839,0.659)  RGB{N0f8}(0.333,0.541,0.157)  RGB{N0f8}(0.388,0.835,0.102)

[:, :, 2] =
 RGB{N0f8}(0.263,0.4,0.416)    RGB{N0f8}(0.769,0.035,0.737)  RGB{N0f8}(0.667,0.545,0.427)
 RGB{N0f8}(0.941,0.612,0.682)  RGB{N0f8}(0.263,0.075,0.278)  RGB{N0f8}(0.396,0.086,0.012)
 RGB{N0f8}(0.325,0.353,0.165)  RGB{N0f8}(0.792,0.114,0.09)   RGB{N0f8}(0.878,0.553,0.537)

[:, :, 3] =
 RGB{N0f8}(0.831,0.29,0.212)   RGB{N0f8}(0.247,0.255,0.384)  RGB{N0f8}(0.655,0.784,0.282)
 RGB{N0f8}(0.275,0.541,0.49)   RGB{N0f8}(0.702,0.894,0.243)  RGB{N0f8}(0.024,0.2,0.804)  
 RGB{N0f8}(0.094,0.294,0.235)  RGB{N0f8}(0.569,0.894,0.714)  RGB{N0f8}(0.455,0.082,0.145)
```
JuliaImages by default says "1 array element" = "1 pixel", which makes writing generic algorithms far easier. We handle color by providing RGB objects; if the NRRD file is annotated to indicate that one axis corresponds to color, that axis gets subsumed into the Color type, which is why this is a 3 and not 4-dimensional array.

In Matlab, I always found it deeply annoying that you could never be certain whether an m-by-n-by-3 array was a 3d grayscale image or a 2d RGB image.

> What happens to this field if we slice the array along the color axis to get out the green channel (should sample units be updated to "gray 1.0"?).

Continuing the demo:
```julia
julia> using Images

julia> map(green, img)
3-dimensional AxisArray{N0f8,3,...} with axes:
    :space_1, 0:2
    :space_2, 0:2
    :time, 0:2
And data, a 3×3×3 Array{N0f8,3}:
[:, :, 1] =
 0.075N0f8  0.173N0f8  0.769N0f8
 0.408N0f8  0.122N0f8  0.886N0f8
 0.839N0f8  0.541N0f8  0.835N0f8

[:, :, 2] =
 0.4N0f8    0.035N0f8  0.545N0f8
 0.612N0f8  0.075N0f8  0.086N0f8
 0.353N0f8  0.114N0f8  0.553N0f8

[:, :, 3] =
 0.29N0f8   0.255N0f8  0.784N0f8
 0.541N0f8  0.894N0f8  0.2N0f8  
 0.294N0f8  0.894N0f8  0.082N0f8
```

Or you could do this:
```julia
julia> imgc = channelview(img)
4-dimensional AxisArray{N0f8,4,...} with axes:
    :color, Base.OneTo(3)
    :space_1, 0:2
    :space_2, 0:2
    :time, 0:2
And data, a 3×3×3×3 Array{N0f8,4}:
[:, :, 1, 1] =
 0.612N0f8  0.855N0f8  0.369N0f8
 0.075N0f8  0.408N0f8  0.839N0f8
 0.302N0f8  0.722N0f8  0.659N0f8

[:, :, 2, 1] =
 0.114N0f8  0.337N0f8  0.333N0f8
 0.173N0f8  0.122N0f8  0.541N0f8
 0.533N0f8  0.118N0f8  0.157N0f8

[:, :, 3, 1] =
 0.263N0f8  0.863N0f8  0.388N0f8
 0.769N0f8  0.886N0f8  0.835N0f8
 0.396N0f8  0.459N0f8  0.102N0f8

[:, :, 1, 2] =
 0.263N0f8  0.941N0f8  0.325N0f8
 0.4N0f8    0.612N0f8  0.353N0f8
 0.416N0f8  0.682N0f8  0.165N0f8

[:, :, 2, 2] =
 0.769N0f8  0.263N0f8  0.792N0f8
 0.035N0f8  0.075N0f8  0.114N0f8
 0.737N0f8  0.278N0f8  0.09N0f8 

[:, :, 3, 2] =
 0.667N0f8  0.396N0f8  0.878N0f8
 0.545N0f8  0.086N0f8  0.553N0f8
 0.427N0f8  0.012N0f8  0.537N0f8

[:, :, 1, 3] =
 0.831N0f8  0.275N0f8  0.094N0f8
 0.29N0f8   0.541N0f8  0.294N0f8
 0.212N0f8  0.49N0f8   0.235N0f8

[:, :, 2, 3] =
 0.247N0f8  0.702N0f8  0.569N0f8
 0.255N0f8  0.894N0f8  0.894N0f8
 0.384N0f8  0.243N0f8  0.714N0f8

[:, :, 3, 3] =
 0.655N0f8  0.024N0f8  0.455N0f8
 0.784N0f8  0.2N0f8    0.082N0f8
 0.282N0f8  0.804N0f8  0.145N0f8
```
`colorview(RGB, imgc)` essentially reverses the transformation. 

> Does it need updating if floating point values are quantized to 16 bits (in which case does it interact with the oddly-named "old min" and "old max" fields?)?

Obviously, in the file it has to reflect however the data are written. Do you mean, if you convert the representation in-memory but would like to re-use the header to write a new file? Overall I'd say that's the job of the programming language/software library/package. In Julia it's trivial because we have a FixedPointNumbers package:

```julia
julia> x = Float32(0.333)
0.333f0

julia> dump(x)
Float32 0.333

julia> y = convert(N0f16, x)
0.333N0f16

julia> dump(y)
FixedPointNumbers.Normed{UInt16,16}
  i: UInt16 21823

julia> 2*y
0.66599524f0
```
(Note that 21823/65535 ≈ 0.333.) We've made the conscious decision to separate the concepts of "meaning" and "representation" when it comes to numbers. We thereby avoid the horrible things that happen in other languages when you convert the element type without also scaling the values.

If you write a Float32 array as NRRD it will be labeled as `single`, and a `N0f16` array will be written as `uint16`, but as long as we know they're intended to represent color or grayscale, then the opposite transformation happens when they are read back in.

**Comment by tommydino93** [comment]
Hi everyone!
Sorry if I intervene in the middle of your very interesting debate, but since the topic is quite related to the first part of the discussion I wanted to kindly ask you something:

I am also having problem loading some .nrrd images in ITK. I have some cardiac MRI images which I want to upload with ITK and then mine with PyRadiomics. 

Here is a link to my nrrd image:
https://mega.nz/#!oLQVFYZZ!PNs-pOWlnXWM9Dau5kwrwYWnsdys4QZjka8uZxAh7ek

When I open it in ImageJ, I see that the `space direction` is (0.89843749,0,0) (0,0.89843749,0) and when I try to load it with my PyCharm script I get the following error:

```
Traceback (most recent call last):
  File "D:/Getting_Started_PyCharm/Pyradiomics/Feature_Extraction_New.py", line 14, in <module>
    image = sitk.ReadImage(imageName)
  File "C:\Users\tommaso\PycharmProjects\Prova\venv\lib\site-packages\SimpleITK\SimpleITK.py", line 8332, in ReadImage
    return _SimpleITK.ReadImage(*args)
RuntimeError: Exception thrown in SimpleITK ReadImage: C:\d\VS14-Win64-pkg\SimpleITK-build\ITK\Modules\IO\NRRD\src\itkNrrdImageIO.cxx:249:
itk::ERROR: NrrdImageIO(000002D43454D0A0): ReadImageInformation: Error reading C:\Users\tommaso\Desktop\Prova_PyCharm\87951608.nrrd:
[nrrd] nrrdLoad: trouble reading "C:\Users\tommaso\Desktop\Prova_PyCharm\87951608.nrrd"
[nrrd] nrrdRead: trouble
[nrrd] _nrrdRead: trouble reading NRRD file
[nrrd] _nrrdFormatNRRD_read: trouble parsing space directions info |(0.89843749,0,0) (0,0.89843749,0)|
[nrrd] _nrrdReadNrrdParse_space_directions: trouble getting space vector 1 of 2
[nrrd] _nrrdSpaceVectorParse: space dimension is 2, but seem to have 3 coefficients
```

What am I missing?

Thank you very much in advance. I apologize if the question is too trivial but I am really stuck right now.

Tommaso

**Comment by ihnorton** [comment]
@tommydino93 the error is mismatched dimensions:

> space dimension is 2, but seem to have 3 coefficients

Looks like it might be a [known bug](http://imagej.1557.x6.nabble.com/Fiji-bug-with-io-Nrrd-Writer-td5002603.html) in ImageJ NRRD implementation.

(speaking a bit generally, you are probably best off to ask on the [ITK forum](discourse.itk.org) for ITK issues or the [Slicer forum](discourse.slicer.org) for PyRadiomics issues, unless they have a separate mailing list. Cheers!)

**Comment by tommydino93** [comment]
Thank you very much @ihnorton !

**Comment by kindlmann** [comment]
@ihnorton you beat me to it.  Too bad something is writing busted NRRD headers.  Also, it's too bad that a more meaningful "space" wasn't used for cardiac MR (such as DICOM's LPS), instead of the no-op "space dimension: 2".
This activity on this threads reminds me how dreadfully long ago it was that I last weighed in here. At the risk of over-sharing, I should note that the existential dread of not getting tenure impeded my ability to justify a lot of time on software development (and hence consistently followup on threads like these), but now that that dread has been lifted, my apparent inability to limit how much time teaching takes is impeding my ability to justify a lot of time on software development.  There are some things I could use help with for getting another version of Teem released (one is CMake-related), but I don't know how to solicit the expertise and/or free labor required.  Any suggestions?

**Comment by ihnorton** [comment]
> Any suggestions?

Move teem to github. I get a bit of existential dread when I see a project on sourceforge these days 😄. Github and free CI from Travis/Appveyor really are a dramatic force-multiplier and have a solid track record of helping build community participation (see [this](http://jakevdp.github.io/blog/2012/09/20/why-python-is-the-last/) and, well, [the whole Julia community](https://juliaobserver.com/)). It's not a panacea -- nothing is, of course -- but it really reduces the friction to contribute (yes, I'm concerned about monoculture and yes, we'll all probably have to scramble to something else when the next credit crunch hits and they are forced to stop lighting VC money on fire; but there are solid alternatives and migration tools ready to go if that happens).

(I've done a lot of svn->git migrations so I could help out a bit there if you want -- email me?)

**Comment by timholy** [comment]
Congrats on tenure, @kindlmann! (Sorry, I had a multi-month absence from email & github back in April).

I'd definitely be interested in a validator for NRRD headers, say something that would error when passed a broken header. We have been writing headers that ITK complained about, for example. Every time I take a dive back into the NRRD docs I think I understand its organization a little bit better, but having a validator would definitely ease my mind with regards to whether we're writing valid headers.

One I've wondered about: suppose you want to use the lovely NRRD orientation facilities but also have a non-spatial axis? For example, we may soon have a lot of images where the spatial axes are actually skewed (e.g., one array axis at 60° relative to another), but there's also a temporal axis. Then it seems like you need `space directions`, `spacings`, and `units`, where `space directions` encodes the funky spatial geometry and `spacings` and `units` also encode the separation between adjacent images along the time axis. Based on my recent experience (see #28) I am not optimistic that ITK will like that.

**Comment by kindlmann** [comment]
A late reply is better than never, right?
I may not be fully understanding the question, but among the "spaces" http://teem.sourceforge.net/nrrd/format.html#space
that can be identified are some that include time.  I included these to enable recording how serial acquisitions of Z slices will have the Z slices extended in both Z and time.  So, the "space direction" for the "Z" axis could have coordinates, say, (0,0,1,0.01), in space LPST, while the "space direction" for the time axis could have coords, (0,0,0,1).
Does that cover the scenario you're imaging/imagining?
As for ITK's understanding of any of this- I have no idea.

(btw I was reminded of my non-answering this because I've forced myself into finally learning Julia, this summer, by saying that I'll use it for an upcoming class in the fall; now I'm finally approaching all this with a new personal interest...)

**Comment by timholy** [comment]
> A late reply is better than never, right?

Unquestionably so!

> Does that cover the scenario you're imaging/imagining?

Close. Specifically, in the upcoming "complex" case I can imagine the most accurate statement would be something like

```
space: microscope-xyz-time
space units: "μm" "μm" "μm" "s"
space direction: (0, 0.4, 0, 0) (0.3464, 0, 0.2, 0) (5, 0, 0, 0.01), (0, 0, 0, 0.25)
```

This describes a configuration in which the first (fastest) axis is horizontal and pointed away from the user, the second axis is left/right relative to the user but also at 30° relative to vertical, the third axis is purely left/right spatially but also separated by 1/100s from the previous slice, and the last reflects the fact that we're capturing 4 volumes per second.

**Comment by timholy** [comment]
Glad to hear about the upcoming use of Julia! I would welcome any role you want to play in the package, whether merely as an advisor or developer.

**Comment by kindlmann** [comment]
That's cool. So is the issue that the NRRD format can represent the per-axis meta-data but ITK doesn't know how to deal, or, is it that the NRRD format doesn't quite capture what you want to represent about the way the imaging happens in space-time?

**Comment by timholy** [comment]
I don't know. I've been using ITK/ImageJ as a validator but I am not sure that its interpretation of the NRRD standard is correct. And I am unsure whether a time axis can really count as a "space unit" or "space direction"?

**Comment by kindlmann** [comment]
For this kind of stuff, I wouldn't trust ITK as a validator.  I'd use "unu minmax".

NRRD lets you describe how axes are oriented in a world space (or, whether they have no presence in the world space, like a short length-2 RFP/GFP axis in microscopy data).  Let the dimension of the array be D, and let the dimension of world-space be WD.  Every axis that has some layout in world space has a non-"none" space direction vector, with WD components (if an axis has no layout in world-space its space direction is "none").  When ever you involved the "space direction" or "space origin" fields, you can specify "space units"; there should be WD of them.

To an extent that I'm sure could confuse ITK, there is actually no relationship between D and WD.  You can have D < WD (e.g. single 2D DICOM image, with its full 3D orientation info), or D == WD (e.g. typical CT scan), or D > WD (multi-channel volume in 3D, or some projection of a 3D volume onto a 2D world-space).  If you are using "space directions", then you need D values, one per axis, either a length-WD vector or "none".  If you are using "space origin", it has to be a single length-WD vector.  If you are using "space units", then you are giving units for each of the WD coordinates in world-space, regardless of D.

So in your example there is no single time axis.  There is space-time, and your array axes have orientation w.r.t. space time.  "space units" just gives the units of space-time.  The fact that D == WD in your example doesn't induce any particular correspondence between the D array axes, and the WD space-time coordinates.  Breaking per-axis orientation free from the per-world-space-coordinates, in the form of a general homogeneous coordinate transform matrix (space origin is the last column, all the space directions are the previous columns), is one of the things that I think NRRD gets right.

Let me know whether that helps.

**Comment by kindlmann** [comment]
Did my explanation above make sense?

**Comment by jasper-tms** [comment]
Hi @kindlmann, this is digging up old history but just wanted to say thanks for the explanations on this thread. Specifically the recommendation to use `unu minmax filename.nrrd` as the source of truth for whether a nrrd header is valid has proven very useful for determining whether I'm making a mistake or ITK/etc just haven't implemented everything.

> To an extent that I'm sure could confuse ITK, there is actually no relationship between D and WD

Yep, sure does, iiuc to this day there's a check in ITK's NRRD reader to insist that `D == WD`: https://github.com/InsightSoftwareConsortium/ITK/blob/cd6608cfe84fc2ae6888c9b2776733a0f1b07a2b/Modules/IO/NRRD/src/itkNrrdImageIO.cxx#L346-L353 which is quite surprising(/disappointing).

Just to make sure it isn't in fact me doing something stupid, would someone mind confirming that this is what a header is supposed to look like for a two-color xyz dataset (dual color scan on a light sheet microscope)?
```
NRRD0005
# This NRRD file was generated by pynrrd
# on 2024-07-26 11:34:03(GMT).
# Complete NRRD file format specification at:
# http://teem.sourceforge.net/nrrd/format.html
type: uint16
dimension: 4
space dimension: 3
sizes: 512 512 100 2
space directions: (0,0,0.8) (0,0.8,0) (2,0,0) none
endian: little
encoding: raw
space units: "microns" "microns" "microns"
```

Any other advice on nrrd readers that have the stuff necessary for a light microscopist to be happy (xyzc and xyzct support) would be appreciated - looks like maybe I should try out Julia? :wink:

Alternatively maybe I'll look into what it would take to expand the ITK nrrd reader myself. (Ultimately I just want to get my data into napari efficiently, and napari uses itk for reading nrrd files.)

**Comment by timholy** [comment]
> looks like maybe I should try out Julia? 😉

That's always the right answer 😆. Seriously, python has an awesome ecosystem (respect), but don't underestimate what JuliaImages might offer you:
- nearly all methods support arbitrary-dimensional arrays
- we have pervasive support for memory-mapped arrays (huge files), and you don't have to "instantiate" them as dense arrays to call methods on them (the dask/scikit-images ecosystem is not nearly as nice as ours in that respect)
- I believe it's fair to say we provide unmatched support for scientific cameras with weird pixel depths. For example, if you have a 14-bit camera, it saturates at what you probably think of as `0x3fff`. We have an unconventional but very general solution for that problem: https://juliaimages.org/latest/tutorials/arrays_colors/#fixedpoint. In biology 16-bit cameras have become more common than they were (I cut my teeth on a 12-bit camera), but the astronomers have some really funky ones like 18-bit and 24-bit so this is still very relevant for scientific imaging.

ImageView was a good viewer before napari even existed, but napari has blown past ImageView with lots of development energy. But Makie is a better plotting ecosystem than anything python has, so 🤷 ? But you'll miss some things about napari.

**Comment by kindlmann** [comment]
@jasper-tms The header you shared looks good to me.  As for how ITK handles NRRD - the code you highlighted (enforcing D == WD) is really a reflection of that point in time when I was motivated to make the NRRD format be part of the NA-MIC world that funded ITK work ~20 years ago, but I wasn't then (and certainly am not now) in a position to push the envelope on how ITK represents image data.  That is the hard bridging that itkNrrdImageIO.cxx tries to do - between the flexibility of the NRRD format and ITK's run-time representation of image data, combined with the assumptions about that run-time representation built into other adjacent ITK-using code.  Maybe it can now handle D != WD in ways that play well with other ITK-using software, more than it could back then, but since I've never actually used ITK for any work, or any other ITK-using software, I wouldn't be in a good position to give tips on how to fundamentally expand how itkNrrdImageIO.cxx works.  Godspeed!

@timholy Having just working on some new bridging between Python (using CFFI) and Teem (the C code base that contains my Nrrd implementation, as well as my other code), I'm looking forward this summer to seeing how much easier it will be in Julia.

**Comment by timholy** [comment]
Do note that we implement NRRD in pure Julia. If you wanted to wrap Teem (which would be quite nice, as it would provide the reference ground-truth), in certain respects Julia and Python would be similar: both languages offer high-quality C FFIs and the issues of library initialization, resource cleanup, etc. would surely be similar.

But Julia might have two advantages:
- thanks to [BinaryBuilder](https://github.com/JuliaPackaging/BinaryBuilder.jl) and Julia's package management system in general, Julia packages "just work." That's not been my experience with Python, where I have to steel myself every time I try installing something that's even slightly complicated.
- once you have the basics of the wrapper off the ground, Julia's [AbstractArray interface](https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array) would mean that you could make Teem-backed arrays first-class citizens, and they would interoperate seamlessly with all the existing JuliaImages functionality.

**Comment by timholy** [comment]
Of course, we'd also welcome any contributions you want to make to this implementation; if we got some things wrong, I'd be happy to make a breaking release that has the correct behavior.

**Comment by termi-official** [comment]
Thanks for the elaborations in this thread. They are really helpful when it comes down to understanding some details in the NRRD format.

Before starting to make PRs I have some questions. I use NRRD files to exchange my data between different toolchains. The data comes from different sources and hence I have different coordinate systems associated with them. The coordinate systems are also not properly aligned and hence my "space directions" fields potentially have different non-zero components for each NRRD file. However, I cannot find a good way to add this additional space direction information into the AxisArray used here. Is there any intended way to track the associated coordinate system info?

Related to this, if I understand the standard correctly, then we can have translation without specifying the concrete orientation as in the standard (https://teem.sourceforge.net/nrrd/format.html#spaceorigin)
> This field conveys the same information as the "Image Position" (0020,0032) field of a DICOM file, except the space in which the vector coordinates are given need not be the DICOM-specific LPS space.

However, here https://github.com/JuliaIO/NRRD.jl/blob/8e5e564c2460503e20d119520b7bfa04270d9a74/src/NRRD.jl#L417-L419 parsing the offset is skipped, even if it is present. Hence, the `specifyorientation &&` should be dropped here, right?
